src/data/set/lattice.lean

```diff
@@ -525,6 +532,25 @@ begin
     exact ⟨t, ⟨⟨s, ⟨sS, tTs⟩⟩, xt⟩⟩ }
 end
 
+lemma Union_range_eq_sUnion {α β : Type*} (C : set (set α))
+  {f : ∀(s : C), β → s} (hf : ∀(s : C), surjective (f s)) :
+  (⋃(y : β), range (λ(s : C), (f s y).val)) = ⋃₀ C :=
```

Comment:
This one and the one after it would be better as a rewrite lemma the other way around. This way it's never going to match.

Reply:
I'm not sure I understand. This is the direction in which it simplifies, right? Isn't that usually the direction equations are written in? Also, the only time I use this lemma is rewriting with it from the (current) left to right (to be fair, I wrote this lemma to be applicable there).

Reply:
I'm basing my guess on the other lemmas like `Union_eq_sUnion`, `sUnion_eq_bUnion` and other permutations. The pattern is "any Union can be written as a special case of sUnion", and so it's easier to rewrite this from "simple to complex" because it will match any expression of the form "U x, ..." and produce something more complicated, while rewriting in the reverse direction requires that all the complicated bits exactly match the original pattern - and it may also involve some higher order matching. In this case it's debatable because you have this auxiliary function `f` that you want lean to guess, and I guess the matching problem on the right doesn't get that bad.

But there are some suspicious bits here, like the fact that `f` has domain `C`, which is a set, so there are coes in a few places; and this lemma is combining 3 or 4 different operations together, in a way that makes me suspect it will have limited utility unless you specifically craft the application to fit the lemma.